Drop node selectors and use ingress [skip ci]

fabiand opened the pull request:
This patch set does two things:

1. Drop a lot of `nodeSelectors`
2. Drop `externalIPs` from services and move inbound access to ingress

Dropping the *nodeSelectors` is done in order to be independent of hostnames.
Dropping the `externalIPs` from services is done to make the services independent of the external setup (i.e. the external IP). Instead all external access is now driven by an ngninx ingress controller.

These changes are mainly driven by the requirement to use the manifests in other cluster setups, or in other words, to make the manifests more generic.

fabiand commented:
This patch set is getting us a bit closer of being able to reuse the manifests easily - or even use them to compile manifests for releases.

Take a look at #265 to see what is still missing to make them generic (or do `grep {{ manifests/*`)

fabiand commented:
```
Ran 41 of 47 Specs in 459.220 seconds
SUCCESS! -- 41 Passed | 0 Failed | 0 Pending | 6 Skipped PASS
ok  	kubevirt.io/kubevirt/tests	459.319s
```

fabiand commented:
ci retest this please

fabiand commented:
retest this please

fabiand commented:
Okay, previos tests failed on CI, becaus the wrong tip got tested. Is settled now:

```
Ran 41 of 47 Specs in 533.693 seconds
SUCCESS! -- 41 Passed | 0 Failed | 0 Pending | 6 Skipped PASS
ok  	kubevirt.io/kubevirt/tests	533.853s
$ git describe --always
v0.0.3-131-gca1b03d
```

fabiand commented:
retest this please

cynepco3hahue commented on cluster/vagrant/setup_kubernetes_master.sh:
```diff
@@ -50,6 +50,35 @@ else
+
+  # Pin the controller to the master
+  ( curl -L https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/with-rbac.yaml && \
+    echo -e "      nodeSelector:\n        kubernetes.io/hostname: master" \
```
Can we export nodeSelector from a config file instead of hardcode it?

fabiand replied:
Actually - I think we can drop this pinning in general

fabiand commented:
@cynepco3hahue it should actually simplify the deployment.

Previously you had to touch all services to adjust them to the cluster.
Now you can you a single ingress configuration to specify how you want to expose KubeVirt services on your cluster.

The ingress will actually also allow us to drop the squid proxy, as ingress-nginx  might be able to do the same.

cynepco3hahue commented:
@fabiand Yes, but ingress controller deployment includes a lot of different manifests, so this configuration not so trivial, by the way, we can install it from charts https://github.com/kubernetes/charts/tree/master/stable/nginx-ingress

kfox1111 commented:
I suggest keeping node selectors, but doing it less selectively. I run non uniform clusters in production, so everything I launch need a node selector to target a pool of nodes appropriate for that service.

Maybe something like nodeselector: {kubevirt-squid: v1}

Ideally though the node selectors would be overridable. this might be a good use case for making a helm chart out of the yaml's.

davidvossel commented on hack/build-manifests.sh:
```diff
@@ -35,7 +35,6 @@ for arg in $args; do
         -e "s/{{ primary_node_name }}/$primary_node_name/g" \
-        -e "s/{{ docker_tag }}/$docker_tag/g" \
-        -e "s/{{ docker_prefix }}/$docker_prefix/g" \
+        -e "/image:/ s#\(.*image:\).*/\(.*\):.*#\1 ${docker_prefix}/\2:${docker_tag}#" \
```
I think this is too broad of a search/replace. It could have some confusing consequences in the future. 

Can we limit this to kubevirt related images?

fabiand replied:
Yes - I actually consider dropping this patch in general and use the templating mechanism we had before.

After all the templates themselfes are not an issue (in this patchset), it's rather that I want to drop clster specific things. But the images are not such a thing.

So let me think about it :)

fabiand replied:
Dropped this patch,

davidvossel commented on automation/test.sh:
```diff
@@ -124,7 +124,14 @@ while [ "$(kubectl get pods -o'custom-columns=status:status.containerStatuses[*]
 
-kubectl get pods
+# Wait until all pods are running
+while [ -n "$(kubectl get pods --no-headers --all-namespaces | grep -v Running)" ]; do
```
"Running" doesn't mean the cluster is operational. For example, haproxy can be running but not in a "ready" state to accept traffic. This will cause the functional tests to fail immediately because they can't contact the cluster. 

Also, you might want to ignore "terminating" phase here.

fabiand replied:
Yes, there is room for improvement, but it's a little better than before.

stu-gott commented on manifests/ingress.yaml.in:
```diff
@@ -0,0 +1,44 @@
+#
+# This is two staged:
+# 1. Tell nginx where which potrs should be redirected to
```
typo: potrs

fabiand commented on cluster/vagrant/setup_kubernetes_master.sh:
```diff
@@ -50,6 +50,29 @@ else
+  curl_yaml_combine \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/default-backend.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/configmap.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/tcp-services-configmap.yaml \
```
I just create them because otherwise errors will pop-up in the events.

fabiand commented on cluster/vagrant/setup_kubernetes_master.sh:
```diff
@@ -50,6 +50,29 @@ else
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/default-backend.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/configmap.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/tcp-services-configmap.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/udp-services-configmap.yaml \
```
Errors appear in the logs. Initially I did not deploy it, but added it to prevent the errors.

fabiand commented on manifests/ingress.yaml.in:
```diff
@@ -0,0 +1,44 @@
+  name: ingress-nginx
+  namespace: kube-system
+spec:
+  externalIPs:
```
Eventually. It would be great. I'd love it.
So far I think it's a bug in CNI which is preventing this (https://github.com/kubernetes-incubator/kube-aws/issues/704).

At least NodePort does not work, that's why I went with external IP for now.

I should add a FIXME somewhere so we get back to it …

davidvossel replied:
ugh... okay. we'll need a issue to track getting rid of ExternalIP for good.

fabiand replied:
There you go, #542

fabiand commented on cluster/vagrant/setup_kubernetes_master.sh:
```diff
@@ -50,6 +50,29 @@ else
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/udp-services-configmap.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/rbac.yaml \
+    https://raw.githubusercontent.com/kubernetes/ingress-nginx/master/deploy/with-rbac.yaml \
+      | sed "s#namespace: ingress-nginx#namespace: kube-system#" \
```
I'd liekt o not do this now - The benefit is that we inheirt whatever changes there.

Once it becomes an issue, then I'll be the first to take a look how to improve it.

fabiand commented:
Pushed a refreshed patchset with the following changes:

- Fixed potrs
- Removed the image regexp and manifest renaming (dropping .in if not a template)

fabiand commented:
Before the refresh it passed actually:

```
Ran 47 of 47 Specs in 867.054 seconds
SUCCESS! -- 47 Passed | 0 Failed | 0 Pending | 0 Skipped PASS
ok  	kubevirt.io/kubevirt/tests	867.107s
```

It currently fails on our CI, because this CI is not rebuilding the vagrant images. The oVirt CI however is doing it IIUIC, thus it should pass at some point.

davidvossel commented on cluster/vagrant/setup_kubernetes_master.sh:
```diff
@@ -50,6 +50,29 @@ else
+  curl_yaml_combine() { for URL in $@; do curl -L $URL ; echo --- ; done }
+
+  # Do the basic deployment
+  curl_yaml_combine \
```
weave's manifest is versioned. I'm trying to avoid a situation where our tests all start breaking because these ingress manifests change. When that happens we waste a day trying to unblock everyone rather than updating on our own terms.

If there's no version number in there, we really don't have any guarantees about stability.

With weave we're at least using something that resembles a release. 
```
kubectl apply -f https://github.com/weaveworks/weave/releases/download/v1.9.4/weave-daemonset-k8s-1.6.yaml
```

fabiand replied:
Ah good point. ket me try something

fabiand replied:
I can't use a previous tag, as there were significant changes. But there will be a release before the end fo week (let's see in what TZ), then I will use versioned files.

fabiand commented:
ci retest this please

fabiand commented:
> Just write your manifests which are tailored to your needs there.

This is precisely why I want to generalize the stock manifests a little more, in order to be able to avoid that everybody has to write their own manifests - after all they are not trivial, and they need to stay in code with the releases.

rmohr commented:
> This is precisely why I want to generalize the stock manifests a little more, in order to be able to avoid that everybody has to write their own manifests - after all they are not trivial, and they need to stay in code with the releases.

And I agree that having recommendations and manifests for that case can make sense, but these manifests are not the one. See all the reasons above.

fabiand commented:
Also, the use of ingress is primarily to separate the internal services from public exposure.
LB is something nice, but not what this is about.

rmohr commented:
> Also, the use of ingress is primarily to separate the internal services from public exposure.
> LB is something nice, but not what this is about.

An explicit public service can act as a separator too. I don't see a benefit for us here.

rmohr commented:
> Also, the use of ingress is primarily to separate the internal services from public exposure.
> LB is something nice, but not what this is about.

And if you go through my spice proxy comments above, you will also see that we don't gain anything  from any kind of stock ingress controller compared to a public service for spice per se.

fabiand commented:
So there is not a lot happening with the manifests themselves, except that the nodeSelectors are dropped and the externalIPs, and because nothing else is changed I am not sure why the change is such an issue.
Especially because oyu say these manifests are for testing, and the testing is not impacted by this change at all.

You could say that an ingress controller is a new dependency, but we see that it's part of many Kubernetes cluster anyway (minikube, openshift, tectonic,..) so I consider it to be a default (but cluster setup specific) component.

The generalization does also not prevent people from pinning certain pods to certain nodes, they can use `kubectl patch` for this - or we use these generalized manifests to consume them in a playbook and make these parts configurable. But the first step is the generalize them _a little_ before we can parametrize them.

> An explicit public service can act as a separator too. I don't see a benefit for us here.

Sure, but the issue is that in that case, you need to mix external informations (external IP) with internal requirements (the service needs to be discoverbale within the cluster [true for some of our services]).
Also - You are not just mixing internal and external informations - you also need to touch internal component configurations to expose them to the public. The ingress concept allows you to separate this.

In the end I do not understand the big objection.

fabiand commented:
ci retest this please

rmohr commented:
> In the end I do not understand the big objection.

I am not willing to bind these manifests to external uses (e.g. playbooks, ...). It is almost impossible for us to then change them. Furhter I wanted to highlight that while ingress is nice, and one can use it if he likes, we don't need it and it brings extra complexity for the core.

fabiand commented:
So, the idea is to publish them elsewhere after a release - then we have release specific manifests to use. This gives us all the freedom to change them between releases.

Btw I dropped the "sane image name" patch previously, becuase I also disliked that change.

rmohr commented:
> In the end I do not understand the big objection.

Actually I am surprised that you didn't expect that. Since I raised the same objections multiple times offline and I am not making them up just for fun. However, there are multiple maintainers. If they see it differently then be it.

rmohr commented:
@fabiand skipping CI for now. We will need to push an extra change which makes sure that we remove ingress related stuff on redeploy.

davidvossel commented:
@fabiand @rmohr  I fear this conversation is in danger of becoming more complex than it needs to be. I believe this PR is an important step forward for us. Lets sort through what's going on here so we can establish some common ground and gain a clear understanding of what's going on here. 

1. **Our use of Ingress in this patch series is specifically for our vagrant setup.** I would never use Ingress like this in production. Most people aren't exposing the k8s api over ingress, that would be insane. 

2. **Ingress is not required for KubeVirt to work**. It's just required for our tests to work in our vagrant environment with generic manifests that don't have hard coded IPs. In the real world people bounce into a bastion host within the cluster (or within the cluster's local network) to invoke the k8s api.

3. **Spice access external to k8s cluster network is a separate problem** We don't have to solve this for production right now. I don't think anyone is considering external spice access solved by nginx ingress. Nothing has changed in that regard.

4. **With this change, our dev manifests don't have to be maintained for every use case**  These are still dev manifests and are subject to change. They aren't versioned. There's no promises being made here. This simply lowers the friction points for people who want to experiment with KubeVirt outside of our vagrant setup. 

5. **node selectors aren't necessary**  Our apiserver problem that requires special placement will go away soon and is a completely separate issue from ingress.  Node selectors shouldn't be a requirement to deploy KubeVirt. If people deploying KubeVirt are opinionated about placement, they can use taints/tolerances for the kubevirt infrastructure. 

Now, what all that said, all Fabian has done here is move the use of hard coded IPs out of the KubeVirt component manifests, and centralized all that into the ingress controllers. He's provided a logical separation between what it means to run our components in a k8s cluster, and how those components are consumed by external networks.

This is an important first step for us. It doesn't solve all our problems, but it breaks the problem up correctly.

While the our dev setup will differ in many ways from how kubevirt is deployed in production, I expect the internal services and KubeVirt component manifests to very closely resemble one another.

rmohr commented:
> Now, what all that said, all Fabian has done here is move the use of hard coded IPs out of the KubeVirt component manifests, and centralized all that into the ingress controllers. He's provided a logical separation between what it means to run our components in a k8s cluster, and how those components are consumed by external networks.

Regarding to ingress: We could also centralise all public services in one file, if that is the progress here. I prefer to have the external IP two or three times in the template, compared to introduce the whole ingress service. I agree that it looks slightly cleaner, to only have to specify the exposed IP once (also service ports instead of ingress is ok for me), but it is one more component which needs to work. Further, for spice the ingress controller can't be used. So what do we gain here? A whole new component just  to remove one or two template entries?

Regarding the node selector: Right now some tests rely on the components to run on the master node. That is probably also one of the reasons, why the tests succeed for @fabiand on the one node setup, but not on CI. That can be changed. If everything works fine, we can remove them.

> While the our dev setup will differ in many ways from how kubevirt is deployed in production, I expect the internal services and KubeVirt component manifests to very closely resemble one another.

Right, therefore it is tempting to "re-use" them. I would prefer to make it clear, that they are not there to be used outside of the core repo (without being reviewed and adjusted at least). I think it is important to have manifests for the most common production setups, but I want them to be separate (these can also live and be tested inside the core repo, if necessary).

davidvossel commented:
> Regarding to ingress: We could also centralise all public services in one file, if that is the progress here. I prefer to have the external IP two or three times in the template, compared to introduce the whole ingress service.

@fabiand @rmohr I think this is a less invasive way to achieve what we want without an ingress controller. By separating out the public services from the KubeVirt component manifests we get a logical separation between the KubeVirt components and ingress without having to introduce usage of the ingress controller. 

The public services that expose the external IP could just become some "vagrant-services.yaml" manifest. The kubevirt component manifests would then remain clean from an ingress perspective. 

We still have the issue of the node selectors being in the kubevirt components. I suggest we remove haproxy and virt-api. The spice and console api's could just become their own separate service endpoint for now. When we move on to the aggregated api server, spice/console could then seamlessly get rolled into that service. 

This involves three parts.
* move our External IP vagrant services out of kubevirt component manifests into vagrant-services.yaml
* get rid of HA proxy and move Spice/Console rest api into it's own service endpoint. 
* remove node selectors from kubevirt manifests. 

 How do you all feel about this for a plan? This gives us generic kubevirt component manifests and doesn't depend/involve an ingress controller.

rmohr commented:
> The spice and console api's could just become their own separate service endpoint for now. When we move on to the aggregated api server, spice/console could then seamlessly get rolled into that service.

I just checked how the apiserver proxying is done. So if we register a new group, e.g. `subresources.kubevirt.io` via an aggregated apiserver, everything sent to `subresources,kubevirt.io/v1alhpa1/...` would be redirected to that aggregated apiserver, which can be stateless, just serves the subresources for the VMs, and we are covered by the k8s authentication/authorization, which is a huge win. Later we just have to remove the `subresources` from the group name and ship a combined aggregated apiserver.

rmohr commented:
Since this means sticking to CRDs for now, we should also add the openapi specification to the CRDs and and start using initializers for basic defaulting (simple defaulting, not related to devices, ...) and additional validations.

davidvossel commented:
I've created issues to track everything we've discussed here. 

#553 - remove haproxy and replace with a stateless aggregated api server (for console/spice)
#554 - separate ingress services from kubevirt components
#555 - remove node selectors (once stateless aggregated api server is done)
#556 - use k8s dynamic intializers and admission controllers for cluster object validation and defaulting.

fabiand commented:
@davidvossel thanks for creating those tickets.

I'll close this PR in favor of #560.
